zcash_primitives/src/zip32.rs

```
@@ -43,41 +48,148 @@ impl ConditionallySelectable for AccountId {
+    }
+
+    fn hardened(index: u32) -> Self {
+        Self(FlexibleKind::Hardened(index))
+    }
+
+    fn is_hardened(&self) -> bool {
+        matches!(self.0, FlexibleKind::Hardened(_))
+    }
+
+    fn index(&self) -> u32 {
+        match self.0 {
+            FlexibleKind::NonHardened(i) => i,
+            FlexibleKind::Hardened(i) => i,
         }
     }
+}
 
-    fn master() -> Self {
-        ChildIndex::from_index(0)
+impl Hardening for Flexible {}
+
+/// A child index for a derived key.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct ChildIndex<H: Hardening>(H);
+
+impl<H: Hardening> ChildIndex<H> {
+    /// Constructs the sentinel `ChildIndex` to use for master keys (that are not derived
+    /// child keys).
+    pub fn master() -> Self {
+        Self(H::master())
     }
 
-    fn value(&self) -> u32 {
-        match *self {
-            ChildIndex::Hardened(i) => i + (1 << 31),
-            ChildIndex::NonHardened(i) => i,
+    /// Constructs a hardened `ChildIndex` from the given value.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `value >= (1 << 31)`.
+    pub fn hardened(value: u32) -> Self {
+        assert!(value < (1 << 31));
+        Self(H::hardened(value + (1 << 31)))
+    }
+
+    /// Returns `true` if this `ChildIndex` is hardened.
+    pub fn is_hardened(&self) -> bool {
+        self.0.is_hardened()
+    }
+
+    /// Returns the index as an integer.
+    pub fn index(&self) -> u32 {
+        self.0.index()
+    }
+}
+
+impl ChildIndex<Flexible> {
+    /// Constructs a `ChildIndex<Flexible>` from the given index.
+    pub fn flexible_from_index(index: u32) -> Self {
+        match index {
+            n if n >= (1 << 31) => Self(Flexible(FlexibleKind::Hardened(n))),
+            n => Self::non_hardened(n),
         }
     }
+
+    /// Constructs a non-hardened `ChildIndex` from the given value.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `value >= (1 << 31)`.
+    pub fn non_hardened(value: u32) -> Self {
```

Why have this panic instead of return `Option<Self>`?